### PR TITLE
Experimental: Set the Tanzu Hub endpoint when creating `tanzu` context

### PR DESCRIPTION
**Pull request by anujc25**
### What this PR does / why we need it
- Set the Tanzu Hub endpoint by fetching it from CSP when creating the `tanzu` context
- The configured endpoint is supposed to be used for creating the Tanzu Hub Client https://github.com/vmware-tanzu/tanzu-plugin-runtime/pull/175

### Which issue(s) this PR fixes

Fixes #

### Describe testing done for PR

```
$ tz login --staging --endpoint https://api.tanzu-dev.cloud.vmware.com
[i] Opening the browser window to complete the login
Log in by visiting this link:

    https://console-stg.cloud.vmware.com/csp/gateway/discovery?client_id=tanzu-cli-client-id&code_challenge=qvaHqJ1OM_170c3Bf5AP6ZkSbSIet8OiqX1RtPWMcx8&code_challenge_method=S256&redirect_uri=http%3A%2F%2F127.0.0.1%3A56866%2Fcallback&response_type=code&state=e9d6d53b690d4539fb13ee975bfe7014

    Optionally, paste your authorization code: [...]


[ok] Successfully logged into 'TAP pre-integration' organization and created a tanzu context
[i] Fetching recommended plugins for active context 'TAP_pre-integration-staging-d03c5c97'...
[ok] No recommended plugins found.
```

```
$ cat ~/.config/tanzu/config-ng.yaml
...
    - name: TAP_pre-integration-staging-d03c5c97
      target: tanzu
      contextType: tanzu
      globalOpts:
        endpoint: https://api.tanzu-dev.cloud.vmware.com
        auth:
            issuer: https://console-stg.cloud.vmware.com/csp/gateway/am/api
            userName: anujc
            permissions:
                - external/39721d32-3962-4a75-83d9-9b3dae23c39d/tap:viewer
                - external/39721d32-3962-4a75-83d9-9b3dae23c39d/tap:admin
                - csp:org_member
                - external/39721d32-3962-4a75-83d9-9b3dae23c39d/tap:member
            accessToken: ...
            IDToken: ...
            refresh_token: ...
            expiration: 2024-04-23T00:08:46.392013-07:00
            type: id-token
      clusterOpts:
        endpoint: https://api.tanzu-dev.cloud.vmware.com/org/ae93ebb4-a249-4553-aa1e-c87c4b7f75e5
        path: /Users/anujc/.kube/config
        context: tanzu-cli-TAP_pre-integration-staging-d03c5c97
      discoverySources: []
      additionalMetadata:
        tanzuHubEndpoint: https://be.symphony-dev.com/hub
        tanzuMissionControlEndpoint: https://tmc.tanzu-dev.cloud.vmware.com
        tanzuOrgID: ae93ebb4-a249-4553-aa1e-c87c4b7f75e5
        tanzuOrgName: TAP pre-integration
...
```

### Release note

```release-note
Experimental: Configure the Tanzu Hub endpoint when creating the `tanzu` context
```

### CODE REVIEW

File: pkg/command/context.go [hunk 1]
```diff
@@ -646,6 +646,15 @@ func globalTanzuLogin(c *configtypes.Context, generateContextNameFunc func(orgNa
 	if err != nil {
 		return err
 	}
+
+	// Fetch the Tanzu Hub endpoint for the Tanzu context as a best case effort
+	tanzuHubEndpoint, err := csp.GetTanzuHubEndpointForTAP(claims.OrgID, c.GlobalOpts.Auth.AccessToken, staging)
+	if err != nil {
+		log.V(7).Infof("unable to get Tanzu Hub endpoint. Error: %v", err.Error())
+	} else {
+		c.AdditionalMetadata[config.TanzuHubEndpointKey] = tanzuHubEndpoint
```
[the remark this answers is not in the document]
Comment (anujc25) [reply pkg/command/context.go:655]: Amm. We only overwrite the metadata if the same context already exists right? So, will this be an issue for new contexts?

[the remark this answers is not in the document]
Comment (vuil) [reply pkg/command/context.go]: can we move this to after L666?

[the remark this answers is not in the document]
Comment (marckhouzam) [reply pkg/command/context.go]: Good point @anujc25. Could the endpoint change on a re-login?

Upgrading an existing context when we first release this feature would be impacted but maybe that’s not a big deal?

**Review by vuil — APPROVED** [review]
lgtm, modulo the go dep update

File: pkg/auth/csp/tanzu.go [hunk 2]
```diff
@@ -427,3 +428,61 @@ func GetOrgNameFromOrgID(orgID, accessToken, issuer string) (string, error) {
...
+	defer resp.Body.Close()
+	body, _ := io.ReadAll(resp.Body)
+	services := CSPServices{}
+
+	if err = json.Unmarshal(body, &services); err != nil {
+		return "", errors.Wrap(err, "could not unmarshal the services response")
+	}
+
+	for _, s := range services.ServicesList {
+		if s.ProductIdentifier == "TANZU-SAAS" && strings.Contains(s.DisplayName, "Tanzu Application Platform") { // TODO: Can this be improved to use some unique id?
```
[the remark this answers is not in the document]
Comment (anujc25) [reply pkg/auth/csp/tanzu.go:480]: Good point. I will create a follow up PR to use Central Configuration Option to fetch this details.

File: pkg/command/context.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (anujc25) [reply pkg/command/context.go]: Sorry, I didn't see the further discussion on this thread. But I have updated the PR to account for re-login with the recent change.

**Pull request merged**